### PR TITLE
Add ability to write findings data to bucket

### PR DESCRIPTION
**Pull request by dav3r**
#  #

## 🗣 Description ##

This PR adds a script (and related resources) that allows an assessor to copy data to the assessment findings bucket.

## 💭 Motivation and context ##

We want to allow data to be copied from COOL assessment environments to the assessment findings bucket.

Related to:
* https://github.com/cisagov/findings-data-import-terraform/pull/34
* https://github.com/cisagov/cool-accounts/pull/132

This is part of the work required for https://github.com/cisagov/cool-system-internal/issues/122.

## 🧪 Testing ##

@jsf9k and I successfully applied this Terraform in a Staging environment (after I applied the Terraform in https://github.com/cisagov/findings-data-import-terraform/pull/34 and https://github.com/cisagov/cool-accounts/pull/132) and we confirmed the following:
* The new `copy-findings-data-to-bucket.sh` script was correctly generated by cloud-init.
* We were able to successfully copy a JSON file to the assessment findings bucket using `copy-findings-data-to-bucket.sh`.
* For assessments that do not require the ability to copy data to the findings bucket, the `findings_data_bucket_name` Terraform variable can be set to an empty string (`""`).

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

- [x] Coordinate with NOM to ensure they know how to update Production tfvars files.

### CODE REVIEW

**Review by jsf9k — APPROVED** [review]

File: cloud-init/write-kali-aws-config.tpl.sh [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+# Now retrieve the VNC user's username from SSM Parameter Store
+ssm_response=$(AWS_ACCESS_KEY_ID=$access_key_id \
+  AWS_SECRET_ACCESS_KEY=$secret_access_key \
+  AWS_SESSION_TOKEN=$session_token \
+  aws --region "${aws_region}" \
+  ssm get-parameter \
+  --name /vnc/username --with-decryption)
+vnc_username=$(sed --quiet \
+  's/^[[:blank:]]*"Value": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$ssm_response")
```
Comment (mcdonnnj) [inline cloud-init/write-kali-aws-config.tpl.sh]: If all we need is the VNC username can't we just take it as a template argument and use the value of https://github.com/cisagov/cool-assessment-terraform/blob/513726d23126a20b11c23dfb70439e1a6d90755e/locals.tf#L44-L48 like we do in https://github.com/cisagov/cool-assessment-terraform/blob/513726d23126a20b11c23dfb70439e1a6d90755e/terraformer_cloud_init.tf#L44? This also moves away from hard-coding the parameter name.

```suggestion
```

File: cloud-init/write-terraformer-aws-config.tpl.sh [hunk 2]
```diff
@@ -4,7 +4,7 @@
 #
```
Comment (mcdonnnj) [inline cloud-init/write-terraformer-aws-config.tpl.sh:4]: Since I'm assuming the VNC username bits above were sourced from this script does it also apply here?

File: kali_cloud_init.tf [hunk 3]
```diff
@@ -62,4 +62,61 @@ data "cloudinit_config" "kali_cloud_init_tasks" {
...
+  # * vnc_username_parameter_name - the name of the SSM Parameter Store
+  #   parameter containing the VNC user's username
+  part {
+    content = templatefile(
+      "${path.module}/cloud-init/write-kali-aws-config.tpl.sh", {
+        aws_region                          = var.aws_region
+        findings_data_bucket_write_role_arn = data.terraform_remote_state.sharedservices.outputs.assessment_findings_write_role.arn
+        permissions                         = "0400"
+        vnc_read_parameter_store_role_arn   = aws_iam_role.guacamole_parameterstorereadonly_role.arn
+        vnc_username_parameter_name         = var.ssm_key_vnc_username
```
Comment (mcdonnnj) [inline kali_cloud_init.tf]: This aligns with my above suggestions.

```suggestion
        vnc_username                        = data.aws_ssm_parameter.vnc_username.value
```

File: cloud-init/write-kali-aws-config.tpl.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+# * aws_region - the AWS region where the roles are to be assumed
+# * findings_data_bucket_write_role_arn - the ARN of the IAM role that
+#   can be assumed to write to the findings data S3 bucket
+# * permissions - the octal permissions to assign the AWS
+#   configuration
+# * vnc_read_parameter_store_role_arn - the ARN of the role that
+#   grants read-only access to certain VNC-related SSM Parameter Store
+#   parameters, including the VNC username
+# * vnc_username_parameter_name - the name of the SSM Parameter Store
+#   parameter containing the VNC user's username
```
Comment (mcdonnnj) [inline cloud-init/write-kali-aws-config.tpl.sh]:
```suggestion
# * vnc_username - the VNC user's username
```

File: cloud-init/write-copy-findings-data-to-bucket.tpl.sh [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+# Input variables are:
+# * aws_region - the AWS region where the roles are to be assumed
+# * findings_data_bucket_name - the name of the findings data S3
+#   bucket
+# * permissions - the octal permissions to assign the script
+# * vnc_read_parameter_store_role_arn - the ARN of the role that
+#   grants read-only access to certain VNC-related SSM Parameter Store
+#   parameters, including the VNC username
+# * vnc_username_parameter_name - the name of the SSM Parameter Store
+#   parameter containing the VNC user's username
```
Comment (mcdonnnj) [inline cloud-init/write-copy-findings-data-to-bucket.tpl.sh]:
```suggestion
# * vnc_username - the VNC user's username
```

File: cloud-init/write-copy-findings-data-to-bucket.tpl.sh [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+# Now retrieve the VNC user's username from SSM Parameter Store
+ssm_response=$(AWS_ACCESS_KEY_ID=$access_key_id \
+  AWS_SECRET_ACCESS_KEY=$secret_access_key \
+  AWS_SESSION_TOKEN=$session_token \
+  aws --region "${aws_region}" \
+  ssm get-parameter \
+  --name /vnc/username --with-decryption)
+vnc_username=$(sed --quiet \
+  's/^[[:blank:]]*"Value": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$ssm_response")
```
Comment (mcdonnnj) [inline cloud-init/write-copy-findings-data-to-bucket.tpl.sh]:
```suggestion
```

File: kali_cloud_init.tf [hunk 7]
```diff
@@ -62,4 +62,61 @@ data "cloudinit_config" "kali_cloud_init_tasks" {
...
+  # * vnc_username_parameter_name - the name of the SSM Parameter Store
+  #   parameter containing the VNC user's username
+  part {
+    content = templatefile(
+      "${path.module}/cloud-init/write-copy-findings-data-to-bucket.tpl.sh", {
+        aws_region                        = var.aws_region
+        findings_data_bucket_name         = var.findings_data_bucket_name
+        permissions                       = "0500"
+        vnc_read_parameter_store_role_arn = aws_iam_role.guacamole_parameterstorereadonly_role.arn
+        vnc_username_parameter_name       = var.ssm_key_vnc_username
```
Comment (mcdonnnj) [inline kali_cloud_init.tf]:
```suggestion
        vnc_username                      = data.aws_ssm_parameter.vnc_username.value
```

File: kali_iam.tf [hunk 8]
```diff
@@ -40,3 +48,23 @@ resource "aws_iam_role_policy_attachment" "efs_mount_policy_attachment_kali" {
...
+# function.
+data "aws_iam_policy_document" "kali_assume_delegated_role_policy_doc" {
+  statement {
+    actions = [
+      "sts:AssumeRole",
+      "sts:TagSession",
+    ]
+    effect = "Allow"
+    resources = [
+      aws_iam_role.guacamole_parameterstorereadonly_role.arn,
```
Comment (mcdonnnj) [inline kali_iam.tf]: If the VNC username is no longer pulled from SSM I don't think we need to give this permission anymore.

```suggestion
```

File: terraformer_cloud_init.tf [hunk 9]
```diff
@@ -92,18 +92,18 @@ data "cloudinit_config" "terraformer_cloud_init_tasks" {
...
-      "${path.module}/cloud-init/write-aws-config.tpl.sh", {
+      "${path.module}/cloud-init/write-terraformer-aws-config.tpl.sh", {
         assessor_account_role_arn                     = var.assessor_account_role_arn
         aws_region                                    = var.aws_region
         permissions                                   = "0400"
         read_cool_assessment_terraform_state_role_arn = module.read_terraform_state.role.arn
         organization_read_role_arn                    = data.terraform_remote_state.master.outputs.organizationsreadonly_role.arn
         terraformer_role_arn                          = aws_iam_role.terraformer_role.arn
-        vnc_read_parameter_store_role_arn             = aws_iam_role.gucamole_parameterstorereadonly_role.arn
+        vnc_read_parameter_store_role_arn             = aws_iam_role.guacamole_parameterstorereadonly_role.arn
```
Comment (mcdonnnj) [inline terraformer_cloud_init.tf]: I believe my suggestions to rework the other templates applies here. I'm not sure if you want to break that into another PR if this is true but I wanted to at least mention it since the file's being touched.

File: terraformer_iam.tf [hunk 10]
```diff
@@ -67,7 +67,7 @@ data "aws_iam_policy_document" "terraformer_assume_delegated_role_policy_doc" {
       aws_iam_role.terraformer_role.arn,
       module.read_terraform_state.role.arn,
       data.terraform_remote_state.master.outputs.organizationsreadonly_role.arn,
-      aws_iam_role.gucamole_parameterstorereadonly_role.arn,
+      aws_iam_role.guacamole_parameterstorereadonly_role.arn,
```
Comment (mcdonnnj) [inline terraformer_iam.tf]: If the templating changes work I'm not sure if permission to assume this role is still needed so just flagging.

File: variables.tf [hunk 11]
```diff
@@ -109,6 +121,12 @@ variable "email_sending_domains" {
   }
 }
 
+variable "findings_data_bucket_name" {
```
Comment (mcdonnnj) [inline variables.tf:124]: This is an ugly suggestion that I thought should at least be mentioned: If this being an empty string is an invalid resource should we use the `count` meta-argument hack to only create resources around this variable's use if this variable is not empty? Something like:


```terraform
count = (var.findings_data_bucket_name != "") ? 1 : 0
```

**Review by mcdonnnj — CHANGES_REQUESTED** [review]
So on a first pass I had some somewhat complicated suggestions around the templating and the VNC username.

File: kali_iam.tf (hunk 8 above)
Comment (jsf9k) [reply kali_iam.tf]: Please see commit 7227f12.

File: kali_cloud_init.tf (hunk 7 above)
Comment (jsf9k) [reply kali_cloud_init.tf]: Please see commit 7227f12.

File: cloud-init/write-copy-findings-data-to-bucket.tpl.sh (hunk 6 above)
Comment (jsf9k) [reply cloud-init/write-copy-findings-data-to-bucket.tpl.sh]: Please see commit 7227f12.

File: cloud-init/write-copy-findings-data-to-bucket.tpl.sh (hunk 5 above)
Comment (jsf9k) [reply cloud-init/write-copy-findings-data-to-bucket.tpl.sh]: Please see commit 7227f12.

File: cloud-init/write-kali-aws-config.tpl.sh (hunk 4 above)
Comment (jsf9k) [reply cloud-init/write-kali-aws-config.tpl.sh]: Please see commit 7227f12.

File: kali_cloud_init.tf (hunk 3 above)
Comment (jsf9k) [reply kali_cloud_init.tf]: Please see commit 7227f12.

File: cloud-init/write-kali-aws-config.tpl.sh (hunk 1 above)
Comment (jsf9k) [reply cloud-init/write-kali-aws-config.tpl.sh]: Please see commit 7227f12.

File: cloud-init/write-terraformer-aws-config.tpl.sh (hunk 2 above)
Comment (jsf9k) [reply cloud-init/write-terraformer-aws-config.tpl.sh:4]: Please see commit d9b1e38.

File: terraformer_cloud_init.tf (hunk 9 above)
Comment (jsf9k) [reply terraformer_cloud_init.tf]: Please see commit d9b1e38.

File: terraformer_iam.tf (hunk 10 above)
Comment (jsf9k) [reply terraformer_iam.tf]: Please see commit d9b1e38.

File: variables.tf (hunk 11 above)
Comment (jsf9k) [reply variables.tf:124]: In the case of an empty string, the policy that gets created is accepted as valid by Terraform and AWS IAM.  At the same time, the resource pointed to be the policy is invalid; therefore, there is zero possibility that the policy could accidentally allow access to an existing resource.  So I think this is OK as it is and there is no need to introduce the additional complication.  I'm open to being convinced otherwise, though.

Comment (jsf9k) [reply variables.tf:124]: Please see commit 88dbf4e.

File: guacamole_parameterstore_assume_role_policy_doc.tf [hunk 12]
```diff
@@ -14,6 +14,7 @@ data "aws_iam_policy_document" "guacamole_parameterstore_assume_role_doc" {
       type = "AWS"
       identifiers = [
         aws_iam_role.guacamole_instance_role.arn,
+        aws_iam_role.kali_instance_role.arn,
         aws_iam_role.terraformer_instance_role.arn,
```
Comment (mcdonnnj) [inline guacamole_parameterstore_assume_role_policy_doc.tf]: Missed this with the other suggestions but I think we can do this now.

```suggestion
```

**Review by mcdonnnj — APPROVED** [review]
Seems reasonable enough and LGTM ✔
